### PR TITLE
Fix public API tests

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131390
* __->__ #131386

This PR fixes a bug in `test_correct_module_names` introduced in #130497. It also addresses post-fix test failures in:
* `torch/ao/quantization/__init__.py` - set the correct `__module__` for several public API helpers
* `torch/library.py` - add `register_vmap` to `__all__`
* `torch/nn/attention/flex_attention.py` - make `round_up_to_multiple` private by prepending an underscore
* `torch/storage.py` - introduce `__all__` to avoid `Self` being re-exported as a public API
* `torch/distributed/pipelining/schedules.py` - add `ZeroBubbleAlgorithm` to `__all__`

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks a lot for fixing this!

Suppressing BC-linter as the API was introduced by mistake recently. So we're happy to break BC for it.

File: torch/cuda/gds.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
 import os
 import sys
-from typing import Callable, Optional
+from typing import Callable, List, Optional
 
 import torch
 from torch.types import Storage
 
 
+__all__: List[str] = []
```
Comment (mikaylagawarecki) [inline torch/cuda/gds.py:9]: Thank you! And...you might want to rebase as my PR that added this file got reverted :D https://github.com/pytorch/pytorch/pull/130633

**Comment by justinchuby** [comment]
Can members from typing and typing_extensions be exempted? They seem very common

**Pull request closed without merging**